rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+# RESTful Service Specifications
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Goals
+
+ - Uniformity between data-centric APIs and remote procedure APIs
+ - Use of pre-existing standards/conventions when possible
+ - Minimization of bike-shedding
+
+Standards/Conventions used as inspiration:
+ - [Whitehouse API Standards](https://github.com/WhiteHouse/api-standards)
+  - RESTful URLs
+ - [Spring Data REST](http://docs.spring.io/spring-data/rest/docs/current/reference/html/)
+  - paging conventions
+  - uses [JSON HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08)
+ - [JSend](http://labs.omniti.com/labs/jsend)
+  - API response conventions
+
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
```

Comment:
Is it bike-shedding to standardize on a single major version number representing backwards compatibility?

Personally I think it's pretty important but I could be swayed. To summarize:

- I think point of a version number is so that you can support multiple incompatible versions at one time. Otherwise, the version number will never change, so there is no point to it.
- If you're going to support multiple at one time, that's hard, so you might want to reduce the amount of simultaneously deployed versions to a minimum: the latest release of each major version. Assuming you have some sane deprecation plan in place and your rate of breaking changes is low, you probably restrict yourself to one or two versions at a time with this rule. Otherwise you could have a lot of minor version variations supported which could be a maintainability problem that would could potentially get rid of wholesale by restricting to a single version (ex: v1.1, v1.2, v2.0, v2.1 vs v1 and v2).